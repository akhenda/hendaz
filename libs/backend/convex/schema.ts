import { defineSchema, defineTable } from 'convex/server';
import { v } from 'convex/values';

export const roles = v.union(v.literal('admin'), v.literal('member'));
export const fileTypes = v.union(v.literal('image'), v.literal('csv'), v.literal('pdf'));

export default defineSchema({
  favorites: defineTable({
    fileId: v.id('files'),
    orgId: v.string(),
    userId: v.id('users'),
  }).index('by_userId_orgId_fileId', ['userId', 'orgId', 'fileId']),

  files: defineTable({
    fileId: v.id('_storage'),
    name: v.string(),
    orgId: v.string(),
    shouldDelete: v.optional(v.boolean()),
    type: fileTypes,
    userId: v.id('users'),
  })
    .index('by_orgId', ['orgId'])
    .index('by_shouldDelete', ['shouldDelete']),

  users: defineTable({
    image: v.optional(v.string()),
    name: v.optional(v.string()),
    orgIds: v.array(v.object({ orgId: v.string(), role: roles })),
    tokenIdentifier: v.string(),
  }).index('by_tokenIdentifier', ['tokenIdentifier']),
});
