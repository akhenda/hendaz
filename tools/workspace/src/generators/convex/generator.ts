import {
  addDependenciesToPackageJson,
  formatFiles,
  generateFiles,
  GeneratorCallback,
  getWorkspaceLayout,
  names,
  offsetFromRoot,
  readProjectConfiguration,
  runTasksInSerial,
  Tree,
  updateJson,
  updateProjectConfiguration,
  writeJson,
} from '@nx/devkit';
import * as path from 'path';
import { execSync } from 'child_process';
import { ArrayLiteralExpression } from 'typescript';
import { libraryGenerator } from '@nx/js';
import { Linter, lintProjectGenerator } from '@nx/eslint';
import { vitestGenerator } from '@nx/vite';
import { tsquery } from '@phenomnomnominal/tsquery';

import { addVitestPlugin } from './lib/add-vitest-plugin';
import { addLinterPlugin } from './lib/add-linter-plugin';
import { ConvexGeneratorSchema } from './schema';

import { convexTestVersion, convexVersion, edgeRuntimeVM } from '../../utils/versions';

interface NormalizedSchema extends ConvexGeneratorSchema {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  parsedTags: string[];
}

function normalizeOptions(tree: Tree, options: ConvexGeneratorSchema): NormalizedSchema {
  const name = names(options.name).fileName;
  const projectDirectory = options.directory ? `${names(options.directory).fileName}/${name}` : name;
  const projectName = projectDirectory.replace(new RegExp('/', 'g'), '-');
  const projectRoot = `${getWorkspaceLayout(tree).libsDir}/${projectDirectory}`;
  const parsedTags = options.tags ? options.tags.split(',').map((s) => s.trim()) : ['type:convex', 'domain:server'];

  return { ...options, projectName, projectRoot, projectDirectory, parsedTags };
}

function addDependencies(tree: Tree) {
  const dependencies: Record<string, string> = {};
  const devDependencies: Record<string, string> = {
    convex: convexVersion,
    'convex-test': convexTestVersion,
    '@edge-runtime/vm': edgeRuntimeVM,
  };

  return addDependenciesToPackageJson(tree, dependencies, devDependencies);
}

function addFiles(tree: Tree, options: NormalizedSchema) {
  const templateOptions = {
    ...options,
    ...names(options.name),
    offsetFromRoot: offsetFromRoot(options.projectRoot),
    template: '',
  };

  generateFiles(tree, path.join(__dirname, 'files'), options.projectRoot, templateOptions);
}

function updateEslintJson(tree: Tree, options: NormalizedSchema) {
  updateJson(tree, path.join(options.projectRoot, '.eslintrc.json'), (eslintJson) => {
    eslintJson.ignorePatterns = [
      ...eslintJson.ignorePatterns,
      'convex/_generated/**/*',
      'vite.config.ts',
      'tsconfig.json',
    ];

    eslintJson.overrides[0].rules = {
      ...eslintJson.overrides[0].rules,
      'no-process-env': 'off',
    };

    return eslintJson;
  });
}

function updateTsConfigs(tree: Tree, options: NormalizedSchema) {
  updateJson(tree, path.join(options.projectRoot, 'tsconfig.json'), (tsConfigJson) => {
    const tsConfigExcludes = tsConfigJson.exclude ?? [];

    tsConfigJson.compilerOptions.module = 'ESNext';
    tsConfigJson.exclude = [...tsConfigExcludes, './convex/_generated/**/*'];

    return tsConfigJson;
  });

  updateJson(tree, path.join(options.projectRoot, 'tsconfig.lib.json'), (tsConfigJson) => {
    const tsConfigIncludes = tsConfigJson.include ?? [];
    const tsConfigExcludes = tsConfigJson.exclude ?? [];
    const includes = [...tsConfigIncludes].join('==>').replace(/src/gi, 'convex').split('==>');
    const excludes = [...tsConfigExcludes].join('==>').replace(/src/gi, 'convex').split('==>');

    tsConfigJson.include = Array.from(new Set([...tsConfigIncludes, ...includes]));
    tsConfigJson.exclude = Array.from(new Set([...tsConfigExcludes, ...excludes]));

    return tsConfigJson;
  });

  updateJson(tree, path.join(options.projectRoot, 'tsconfig.spec.json'), (tsConfigJson) => {
    const tsConfigIncludes = tsConfigJson.include ?? [];
    const includes = [...tsConfigIncludes].join('==>').replace(/src/gi, 'convex').split('==>');

    tsConfigJson.include = Array.from(new Set([...tsConfigIncludes, ...includes]));

    return tsConfigJson;
  });
}

function updateConvexJson(tree: Tree, options: NormalizedSchema) {
  writeJson(tree, 'convex.json', {
    functions: `${options.projectRoot}/convex`,
  });
}

function updateViteConfig(tree: Tree, options: NormalizedSchema) {
  const filePath = path.join(options.projectRoot, 'vite.config.ts');
  const fileEntry = tree.read(filePath);
  const contents = fileEntry?.toString() ?? '';

  let newContents = tsquery.replace(contents, 'ArrayLiteralExpression', (node) => {
    const trNode = node as ArrayLiteralExpression;

    if (trNode.getText() === "['src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}']") {
      return trNode.getText().replace(/src/gi, '{src,convex}');
    }

    return null;
  });

  newContents = newContents.replace(
    /environment: 'edge-runtime',/gi,
    "environment: 'edge-runtime',\n\tserver: { deps: { inline: ['convex-test'] } },",
  );

  newContents = newContents.replace(
    /reporters: \['default'\],/gi,
    "reporters: process.env['GITHUB_ACTIONS'] ? ['verbose', 'github-actions'] : ['verbose'],",
  );

  // only write the file if something has changed
  if (newContents !== contents) tree.write(filePath, newContents);
}

function initialiseConvex() {
  const result = execSync('npx convex codegen', { stdio: 'inherit' });

  return result;
}

export async function convexGenerator(tree: Tree, options: ConvexGeneratorSchema) {
  const tasks: GeneratorCallback[] = [];
  const normalizedOptions = normalizeOptions(tree, options);

  await libraryGenerator(tree, {
    name: normalizedOptions.projectName,
    compiler: 'tsc',
    unitTestRunner: 'vitest',
    projectNameAndRootFormat: 'as-provided',
    directory: normalizedOptions.projectRoot,
    setParserOptionsProject: true,
    linter: 'eslint',
    strict: true,
    tags: 'type:convex,domain:server',
  });

  const projectConfiguration = readProjectConfiguration(tree, options.name);

  delete projectConfiguration.targets?.build;

  updateProjectConfiguration(tree, normalizedOptions.projectName, {
    root: projectConfiguration.root,
    targets: {
      ...projectConfiguration.targets,
      dev: {
        executor: 'nx:run-commands',
        options: {
          command: 'convex dev',
        },
      },
    },
  });

  addFiles(tree, normalizedOptions);

  tasks.push(await addVitestPlugin(tree));
  tasks.push(addLinterPlugin(tree));
  tasks.push(addDependencies(tree));

  await lintProjectGenerator(tree, {
    project: normalizedOptions.projectName,
    skipFormat: true,
    linter: Linter.EsLint,
  });

  await vitestGenerator(tree, {
    project: normalizedOptions.projectName,
    coverageProvider: 'v8',
    uiFramework: 'none',
    testEnvironment: 'edge-runtime',
  });

  updateEslintJson(tree, normalizedOptions);
  updateConvexJson(tree, normalizedOptions);
  updateTsConfigs(tree, normalizedOptions);
  updateViteConfig(tree, normalizedOptions);

  await formatFiles(tree);

  return () => {
    runTasksInSerial(...tasks);
    initialiseConvex();
  };
}

export default convexGenerator;
