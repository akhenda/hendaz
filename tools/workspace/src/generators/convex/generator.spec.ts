import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { Tree, readProjectConfiguration } from '@nx/devkit';

import { convexGenerator } from './generator';
import { ConvexGeneratorSchema } from './schema';

describe('convex generator', () => {
  let tree: Tree;
  const options: ConvexGeneratorSchema = { name: 'test' };

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  it('should run successfully', async () => {
    await convexGenerator(tree, options);
    const config = readProjectConfiguration(tree, 'test');
    expect(config).toBeDefined();
  });
});
