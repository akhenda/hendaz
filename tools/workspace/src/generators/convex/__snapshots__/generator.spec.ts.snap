// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convex generator snapshots should generate files 1`] = `
"{
  "functions": "./test/convex"
}
"
`;

exports[`convex generator snapshots should generate files 2`] = `
"{
  "extends": [
    "../.eslintrc.json"
  ],
  "ignorePatterns": [
    "!**/*",
    "convex/_generated/**/*",
    "vite.config.ts",
    "tsconfig.json"
  ],
  "overrides": [
    {
      "files": [
        "*.ts",
        "*.tsx",
        "*.js",
        "*.jsx"
      ],
      "rules": {
        "no-process-env": "off"
      }
    },
    {
      "files": [
        "*.ts",
        "*.tsx"
      ],
      "rules": {}
    },
    {
      "files": [
        "*.js",
        "*.jsx"
      ],
      "rules": {}
    }
  ]
}
"
`;

exports[`convex generator snapshots should generate files 3`] = `
"{
  "extends": "../tsconfig.base.json",
  "compilerOptions": {
    "module": "ESNext",
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "noImplicitOverride": true,
    "noPropertyAccessFromIndexSignature": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true
  },
  "files": [],
  "include": [],
  "references": [
    {
      "path": "./tsconfig.lib.json"
    },
    {
      "path": "./tsconfig.spec.json"
    }
  ],
  "exclude": [
    "./convex/_generated/**/*"
  ]
}
"
`;

exports[`convex generator snapshots should generate files 4`] = `
"{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "../dist/out-tsc",
    "declaration": true,
    "types": [
      "node"
    ]
  },
  "include": [
    "src/**/*.ts",
    "convex/**/*.ts"
  ],
  "exclude": [
    "vite.config.ts",
    "src/**/*.spec.ts",
    "src/**/*.test.ts",
    "convex/**/*.spec.ts",
    "convex/**/*.test.ts"
  ]
}
"
`;

exports[`convex generator snapshots should generate files 5`] = `
"{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "../dist/out-tsc",
    "types": [
      "vitest/globals",
      "vitest/importMeta",
      "vite/client",
      "node",
      "vitest"
    ]
  },
  "include": [
    "vite.config.ts",
    "vitest.config.ts",
    "src/**/*.test.ts",
    "src/**/*.spec.ts",
    "src/**/*.test.tsx",
    "src/**/*.spec.tsx",
    "src/**/*.test.js",
    "src/**/*.spec.js",
    "src/**/*.test.jsx",
    "src/**/*.spec.jsx",
    "src/**/*.d.ts",
    "convex/**/*.test.ts",
    "convex/**/*.spec.ts",
    "convex/**/*.test.tsx",
    "convex/**/*.spec.tsx",
    "convex/**/*.test.js",
    "convex/**/*.spec.js",
    "convex/**/*.test.jsx",
    "convex/**/*.spec.jsx",
    "convex/**/*.d.ts"
  ]
}
"
`;

exports[`convex generator snapshots should generate files 6`] = `
"import { defineConfig } from 'vite';
import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';
export default defineConfig({
    root: __dirname,
    cacheDir: '../node_modules/.vite/test',
    plugins: [nxViteTsPaths(),
    ],
    // Uncomment this if you are using workers. 
    // worker: {
    //  plugins: [ nxViteTsPaths() ],
    // },
    test: {
        '': watch, '': false,
        '': globals, '': true,
        '': environment, '': "jsdom",
        '': include, '': ["src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}"],
        '': reporters, '': ["default"],
        '': coverage, '': { "reportsDirectory": "../coverage/test", "provider": "v8" },
        '"reportsDirectory"': "../coverage/test",
        '"provider"': "v8",
        'globals': true,
        'environment': "edge-runtime",
        'include': ["src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}"],
        'reporters': ["default"],
        'coverage': { "reportsDirectory": "../coverage/test", "provider": "v8" },
    },
});"
`;
